File: python/rob_agent.py
import os
import asyncio
from typing import TypedDict, List, Optional
import typer
from dotenv import load_dotenv

# Load env before imports
load_dotenv()

from langchain_openai import AzureChatOpenAI
from langchain_core.messages import HumanMessage, SystemMessage

# Import services
# Assuming running from 'python' directory or PYTHONPATH is set
try:
    from services.docstore import MockDocStore
    from services.nlp import MockNLPService
except ImportError:
    # fix path for direct execution
    import sys
    sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    from python.services.docstore import MockDocStore
    from python.services.nlp import MockNLPService

# Define State
class AgentState(TypedDict):
    question: str
    search_type: Optional[str]
    documents: List[str]
    answer: Optional[str]

# Nodes
async def analyze_node(state: AgentState):
    print("--- NLP Analyzing ---")
    nlp = MockNLPService()
    search_type = await nlp.analyze(state["question"])
    print(f"Suggested Search: {search_type}")
    return {"search_type": search_type}

async def search_node(state: AgentState):
    print("--- Searching DocStore ---")
    docstore = MockDocStore()
    search_type = state["search_type"]
    question = state["question"]

    if search_type == "keyword":
        docs = await docstore.find_by_keyword(question)
    elif search_type == "similarity":
        docs = await docstore.find_by_similarity(question)
    elif search_type == "relations":
        docs = await docstore.find_by_relations(question)
    else:
        docs = []

    print(f"Found {len(docs)} documents.")
    return {"documents": docs}

async def generate_node(state: AgentState):
    print("--- Generating Answer ---")
    # Check if we have Azure keys, else mock
    if not os.getenv("AZURE_OPENAI_API_KEY"):
        print("No Azure OpenAI key found, using Mock LLM.")
        docs_text = "\n".join(state["documents"])
        answer = f"I found these documents relevant:\n{docs_text}\n\nBased on them, the answer is: [Simulated Answer for '{state['question']}']"
        return {"answer": answer}

    llm = AzureChatOpenAI(
        azure_deployment=os.getenv("AZURE_OPENAI_DEPLOYMENT_NAME"),
        api_version=os.getenv("AZURE_OPENAI_API_VERSION", "2023-05-15"),
    )

    docs_text = "\n".join(state["documents"])
    prompt = f"""
    You are a helpful assistant. Use ONLY the following documents to answer the user's question.

    First, analyze the relevance of the given documents and provide a feedback starting with "Feedback: ".
    Then, compose the final answer starting with "Answer: ".
    If the answer is not in the documents, state that in the feedback and answer "I don't know".

    Documents:
    {docs_text}

    Question: {state["question"]}
    """

    response = await llm.ainvoke([HumanMessage(content=prompt)])
    return {"answer": response.content}

# Graph Construction
from langgraph.graph import StateGraph, END

workflow = StateGraph(AgentState)
workflow.add_node("analyze", analyze_node)
workflow.add_node("search", search_node)
workflow.add_node("generate", generate_node)

workflow.set_entry_point("analyze")
workflow.add_edge("analyze", "search")
workflow.add_edge("search", "generate")
workflow.add_edge("generate", END)

app_graph = workflow.compile()

# CLI
app = typer.Typer()

@app.command()
def ask(question: str):
    """
    Ask Rob a question.
    """
    print(f"Rob is thinking about: {question}")
    try:
        result = asyncio.run(app_graph.ainvoke({"question": question}))
        print("\n--- Answer ---")
        print(result["answer"])
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    app()

File: kotlin/src/main/kotlin/agents/RobAgent.kt
package agents

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.parameters.arguments.argument
import kotlinx.coroutines.runBlocking
import services.MockDocStore
import services.MockNlpService
import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import kotlinx.serialization.json.*

// Simulation of Koog DSL
object Koog {
    fun agent(name: String, block: suspend (String) -> String): Agent = Agent(name, block)
}

class Agent(val name: String, val handler: suspend (String) -> String) {
    suspend fun chat(message: String): String = handler(message)
}

class RobCommand : CliktCommand() {
    val question by argument(help = "The question to ask")

    override fun run() = runBlocking {
        val rob = Koog.agent("Rob") { question ->
            echo("--- NLP Analyzing ---")
            val nlp = MockNlpService()
            val searchType = nlp.analyze(question)
            echo("Suggested Search: $searchType")

            echo("--- Searching DocStore ---")
            val docstore = MockDocStore()
            val docs = when (searchType) {
                "keyword" -> docstore.findByKeyword(question)
                "similarity" -> docstore.findBySimilarity(question)
                "relations" -> docstore.findByRelations(question)
                else -> emptyList()
            }
            echo("Found ${docs.size} documents.")

            echo("--- Generating Answer ---")
            val apiKey = System.getenv("AZURE_OPENAI_API_KEY")
            if (apiKey.isNullOrBlank()) {
                echo("No Azure OpenAI key found, using Mock LLM.")
                val docsText = docs.joinToString("\n")
                "I found these documents relevant:\n$docsText\n\nBased on them, the answer is: [Simulated Answer for '$question']"
            } else {
                callAzureOpenAI(question, docs, apiKey)
            }
        }

        val answer = rob.chat(question)
        echo("\n--- Answer ---")
        echo(answer)
    }

    suspend fun callAzureOpenAI(question: String, docs: List<String>, apiKey: String): String {
        val endpoint = System.getenv("AZURE_OPENAI_ENDPOINT")
        val deployment = System.getenv("AZURE_OPENAI_DEPLOYMENT_NAME")
        val version = System.getenv("AZURE_OPENAI_API_VERSION") ?: "2023-05-15"

        val url = "$endpoint/openai/deployments/$deployment/chat/completions?api-version=$version"

        val client = HttpClient {
            install(ContentNegotiation) {
                json(Json { ignoreUnknownKeys = true })
            }
        }

        val docsText = docs.joinToString("\n")
        val prompt = """
            You are a helpful assistant. Use ONLY the following documents to answer the user's question.

            First, analyze the relevance of the given documents and provide a feedback starting with "Feedback: ".
            Then, compose the final answer starting with "Answer: ".
            If the answer is not in the documents, state that in the feedback and answer "I don't know".

            Documents:
            $docsText

            Question: $question
        """.trimIndent()

        try {
            val response = client.post(url) {
                header("api-key", apiKey)
                contentType(ContentType.Application.Json)
                setBody(buildJsonObject {
                    put("messages", buildJsonArray {
                        add(buildJsonObject {
                            put("role", "user")
                            put("content", prompt)
                        })
                    })
                })
            }
            val body = response.bodyAsText()
            val json = Json.parseToJsonElement(body)
            return json.jsonObject["choices"]?.jsonArray?.get(0)?.jsonObject?.get("message")?.jsonObject?.get("content")?.jsonPrimitive?.content ?: "Error parsing response: $body"
        } catch (e: Exception) {
            return "Error calling LLM: ${e.message}"
        } finally {
            client.close()
        }
    }
}

fun main(args: Array<String>) = RobCommand().main(args)
